cep-recipe-jinja.md

```diff
@@ -0,0 +1,320 @@
+
+The version comparison rules follow those of the `conda` version comparison rules.
+
+## The `hash` variable
+
+`${{ hash }}` is the variant hash and is useful in the build string computation. This used to be `PKG_HASH` in the old recipe format. Since the `hash` variable depends on the variant computation, it is only available in the `build.string` field and is computed after the entire variant computation is finished.
+
+## The `env` object
+
+The `env` object is used to retrieve environment variables and inject them into the recipe. There are two ways to do this:
+
+- `${{ env.get("MY_ENV_VAR") }}` will return the value of the environment variable `MY_ENV_VAR` or throw an error if the environment variable is not set.
+- `${{ env.get_default("MY_ENV_VAR", "default_value") }}` will return the value of the environment variable `MY_ENV_VAR` or `"default_value"` if it is not set.
+
+You can also check for the existence of an environment variable:
+
+- `${{ env.exists("MY_ENV_VAR") }}` will return a boolean `true` if the environment variable `MY_ENV_VAR` is set and `false` otherwise.
+
+## Jinja filters
+
+A feature of `jinja` is called "filters". Filters are functions that can be applied to variables in a template expression.
+
+The syntax for a filter is `{{ variable | filter_name }}`. A filter can also take arguments, such as `... | replace('foo', 'bar')`.
+
+The following Jinja filters are available:
+
+- `replace`: replace a string with another string (e.g. `"{{ 'foo' | replace('oo', 'aa') }}"` will return `"faa"`)
+- `lower`: convert a string to lowercase (e.g. `"{{ 'FOO' | lower }}"` will return `"foo"`)
+- `upper`: convert a string to uppercase (e.g. `"{{ 'foo' | upper }}"` will return `"FOO"`)
+- `int`: convert a string to an integer (e.g. `"{{ '42' | int }}"` will return `42`)
+
+There are more filters available as [documented by MiniJinja](https://docs.rs/minijinja/latest/minijinja/filters/index.html).
+
+> [!NOTE]
+> Should we add all filters from `minijinja` to this spec? Probably ...
+
+### Extra filters for recipes
+
+#### The `version_to_buildstring` filter
+
+- `${{ python | version_to_buildstring }}` converts a version from the variant to a build string (it removes the `.` character and takes only the first two elements of the version).
+For example the following 
+
+```yaml
+context:
+  cuda: "11.2.0"
+
+build:
+  string: ${{ hash }}_cuda${{ cuda_version | version_to_buildstring }}
+```
+
+Would evaluate to a `abc123_cuda112` (assuming the hash was `abc123`).
+
+### Various remarks
+
+#### Inline conditionals with Jinja
+
+The new recipe format allows for inline conditionals with Jinja. If they are falsey, and no `else` branch exists, they will render to an empty string (which is, for example in a list or dictionary, equivalent to a YAML `null`).
+
+When we render the recipe, we filter all values that are `null` from the YAML output. This allows us to write and parse the following easily:
```

Comment:
Instead of saying "we". should it say "the build tool must"? I've seen "we" used in other places in this CEP too. ALso, same question for "us".

Reply:
Agreed!